atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java

```diff
@@ -0,0 +1,145 @@
+            ImmutableWriteOnceDeleteOnceWorkflowConfiguration.builder()
+                    .tableConfiguration(ImmutableTableConfiguration.builder()
+                            .tableName(TABLE_NAME)
+                            .isolationLevel(IsolationLevel.SERIALIZABLE)
+                            .build())
+                    .rateLimit(Double.MAX_VALUE)
+                    .iterationCount(1)
+                    .build();
+
+    private final InteractiveTransactionStore memoryStore = AtlasDbTransactionStore.create(
+            TransactionManagers.createInMemory(ImmutableSet.of()),
+            ImmutableMap.of(
+                    TableReference.createWithEmptyNamespace(TABLE_NAME),
+                    AtlasDbUtils.tableMetadata(IsolationLevel.SERIALIZABLE)));
+
+    @Test
+    public void deleteWhenBuggifyTrue() {
+        Workflow workflow = createWorkflow(AlwaysBuggifyFactory.INSTANCE, i -> i);
+        WorkflowHistory history = workflow.run();
+        WitnessedTransaction transaction = Iterables.getOnlyElement(history.history());
+        WitnessedTransactionAction action = Iterables.getOnlyElement(transaction.actions());
+        assertThat(action).isInstanceOf(WitnessedDeleteTransactionAction.class);
+    }
+
+    @Test
+    public void writeWhenBuggifyFalse() {
+        Workflow workflow = createWorkflow(NoOpBuggifyFactory.INSTANCE, i -> i);
+        WorkflowHistory history = workflow.run();
+        WitnessedTransaction transaction = Iterables.getOnlyElement(history.history());
+        WitnessedTransactionAction action = Iterables.getOnlyElement(transaction.actions());
+        assertThat(action).isInstanceOf(WitnessedWriteTransactionAction.class);
+    }
+
+    @Test
+    public void calculateChanceForDeleteHasExpectedValues() {
+        assertThat(WriteOnceDeleteOnceWorkflows.calculateChanceForDelete(0, 100))
+                .isEqualTo(0.01);
+        assertThat(WriteOnceDeleteOnceWorkflows.calculateChanceForDelete(99, 100))
+                .isEqualTo(1.00);
+    }
+
+    @Test
+    public void writesThenDeletes() {
+        Iterator<Integer> keys = List.of(0, 0, 1, 1).iterator();
+        IntFunction<Integer> rowKeyGenerator = _ignore -> keys.next();
+        BuggifyFactory alternatingBuggify = mock(BuggifyFactory.class);
+        when(alternatingBuggify.maybe(anyDouble()))
+                .thenReturn(
+                        NoOpBuggify.INSTANCE, DefaultBuggify.INSTANCE, NoOpBuggify.INSTANCE, DefaultBuggify.INSTANCE);
+        Workflow workflow = createWorkflow(alternatingBuggify, rowKeyGenerator, 4);
+
+        List<WitnessedTransaction> transactions = workflow.run().history();
+
+        assertThat(transactions.size()).isEqualTo(4);
+        assertWriteTransaction(transactions.get(0), 0);
+        assertDeleteTransaction(transactions.get(1), 0);
+        assertWriteTransaction(transactions.get(2), 1);
+        assertDeleteTransaction(transactions.get(3), 1);
```

Comment:
not actionable: the orchestration feels a bit heavy-handed... but I don't really have a nicer solution.